guia/genericos.md

```diff
@@ -1 +1,511 @@
+
+Além disso, `IntStack` especifica que para esta implementação de `Container`, o `ItemType` apropriado a ser usado é um tipo `Int`. A definição `typealias ItemType = Int` torna o tipo de dados abstrato de `ItemType` no tipo concreto `Int` para esta implementação do protocolo `Container`.
+
+Graças à inferência de tipos de `Swift`, você não precisa necessariamente declarar um tipo concreto `Int` de `ItemType` como parte da definição de `IntStack`. Como `IntStack` obedece a todos os requisitos do protocolo `Container`, Swift pode inferir o `IntType` apropriado a ser usado, simplesmente olhando para o tipo do parâmetro `item` do método `append(_:)` e o tipo devolvido pelo subscrito.
+
+De fato, se você remover a sentença `typealias ItemType = Int` no código acima, tudo ainda vai funcionar, porque é claro qual tipo deveria ser usado para `ItemType`.
+
+Você também pode criar um `Stack` genérico obedecendo ao protocolo:
+
+```swift
+struct Stack<Element>: Container {
+    // implementação original de Stack<Element>
+    var items = [Element]()
+    mutating func push(item: Element) {
+        items.append(item)
+    }
+    mutating func pop() -> Element {
+        return items.removeLast()
+    }
+    // conformidade ao protocolo Container
+    mutating func append(item: Element) {
+        self.push(item)
+    }
+    var count: Int {
+        return items.count
+    }
+    subscript(i: Int) -> Element {
+        return items[i]
+    }
+}
+```
+
+Desta vez, o parâmetro de tipo `Element` é usado como tipo para o parâmetro `item` do método `append(_:)` e como o tipo devolvido pelo subscrito. Swift pode a partir disso inferir qual `Element` apropriado deve ser usado como `ItemType` em um contêiner apropriado.
+
+#### Estendendo o Tipo existente para especificar um Tipo Associado
+
+Você pode estender um tipo para adicionar conformidade para um protocolo, como descrito em [Adicionando conformidade para um protocolo com uma Extensão](./protocolos.md#ProtocolConformanceWithExtension). Isto inclui um protocolo com um tipo associado.
+
+O tipo `Array` de Swift já provê um método `append(_:)`, uma propriedade `count`, e um subscrito com um índice `Int` para recuperar estes elementos. Estas três capacidades coincidem com os requisitos do protocolo `Container`. isto significa que você pode estender `Array` para obedecer ao protocolo `Container` simplesmente declarando que `Array` adota o protocolo. Você pode fazer isso com uma extensão vazia, como descrito em [Declarando Adição à um Protocolo com uma extensão](./protocolos.md#DeclaringProtocolAdoptionWithExtension):
+
+```swift
+extension Array: Container {}
+```
+
+O método `append(_:)` existente em `Array` e o subscrito permitem ao Swift inferir o tipo apropriado para usar com `ItemType`, da mesma forma que com o tipo `Stack` genérico acima. Depois de definir esta extensão, você pode usar qualquer `Array` como um `Container`.
+
+### _Clausulas Onde_
```

Comment:
Não tem só se for um texto sobre música :p
https://pt.wikipedia.org/wiki/Clausula